### DISCUSSION TITLE
possibility of running Tensor mechanics problem (with plasticity) using Explicit solvers?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17678] [Q&A Modules: Solid mechanics] [answered]
Hi all, I've been trying to make my mine simulation to be able to run on explicit solver (to match with Abaqus).
I wonder if MOOSE can do what I want.

So far what I've found is that the only Explicit solver from 'TimeIntegrator System' that works with my problem is `CentralDifference with lumped`.

However, it does not go beyond the 'Static Initialization' step > tho I tried to modify the example problem (elastic) and it works fine, not sure if the problem comes from plasticity.

Could you please have a look at my input file (attached) and please guide me on this problem, I really could not find many examples on Explicit solvers.

Thank you very much.
Traiwit

[ExplicitMOOSE.zip](https://github.com/idaholab/moose/files/6374752/ExplicitMOOSE.zip)

**Comment by Traiwit** [comment]
apparently, the current problem is an instability in the smaller elements (in this case the mine's tunnel), where the larger elements (mine stope) look fine.
![Screenshot from 2021-04-27 09-57-50](https://user-images.githubusercontent.com/79562439/116165412-2c1fec80-a73f-11eb-88e9-1693719fd68a.png)

**Reply by hugary1995** [reply]
Does it run if you remove plasticity? Are you sure your time step size satisfies CFL conditions?

**Reply by Traiwit** [reply]
Hi @hugary1995, 

I could run using both elastic and plastic, esp with very small-time step size, however, at some point, the small elements start to explode (like the figure attached above). Also, it takes so much time when the time step size is small, is there anything that works like `mass-scaling` in MOOSE?

It could run fine if I set `implicit=false` in the InertialForce kernel OR when I set `use_constant_mass = true` under TimeIntegrator/type = CentralDifference/lumped >> HOWEVER, disp_ for these cases remain 0, which is incorrect.

is CFL documented in MOOSE? I tried to follow  [https://mooseframework.inl.gov/modules/tensor_mechanics/CriticalTimeStepMath.html](url) for time step size.

tbh I actually don't know what to do, I tried all the explicit solvers, seem like this is my only option.

Thank you,
Traiwit

**Comment by WilkAndy** [comment]
@bwspenc - could you answer the question about how to do solid mechanics using explicit timestepping, or nominate someone else?  It's been about 5 years since i ran any explicit stuff, and i know MOOSE has moved on since then.

**Reply by WilkAndy** [reply]
Prompting @bwspenc

**Reply by Traiwit** [reply]
Thank you for reminding @WilkAndy, I kinda give up on this idea of explicit, now I want to relax plasticity so the simulation doesn't crash when the initial stress is too high or the softening is too rapid (5e6 to 5e5 from plastic strain 0 to 1%). I probs should make a new forum regarding this.

**Reply by bwspenc** [reply]
Unfortunately, I'm probably not very helpful here. As you've already seen, there is a basic capability, but it is pretty limited -- doesn't work with contact, and the last I heard, it was really slow. If it's not a really short duration impact event, you will likely have better luck with the implicit solver for the time being. We just haven't had much of a driver for developing this by our current programs.

**Reply by LabrosV** [reply]
I am currently using the implicit solver to run some problems, but I'm interested in using the explicit solver to investigate short duration impact events. How basic/limited is the capability? (i.e. I'm looking for an updated manual, if exists.) Currently, there is no need for contact and it's OK if it's not so fast.

**Reply by hugary1995** [reply]
@LabrosV, as @bwspenc said above, we haven't looked at this much by ourselves. I see no obvious issue in using the explicit solver with plasticiticty. We are happy to hear about your experience with it.

**Reply by LabrosV** [reply]
Thank you @hugary1995. I'll take a look and share with you my experience.

**Comment by ngrilli** [comment, marked as answer]
Dear @Traiwit 

I attach an example simulation where you can see the settings I use for dynamics simulations:

[Case2ResidualStress4umDynamic.zip](https://github.com/idaholab/moose/files/6421483/Case2ResidualStress4umDynamic.zip)

My experience is that dynamics simulations in moose work fine without explicit method, just use the PJFNK solver.
Explicit dynamic simulations are useful for impact or other specialized simulations.
If you just want to see the wave dynamics, implicit is fine.

My experience is that convergence problem may arise from:

1) Newmark parameters not set to the stable values, see here:
https://mooseframework.inl.gov/modules/tensor_mechanics/Dynamics.html

2) Time step must be set such that it takes about 20 time steps for a wave to cross the smallest element (calculate wave speed analytically). You need to check the same for the load you are applying, if the load is very fast.

3) use_displaced_mesh = false
in the kernel leads to more stable simulations, but then it's an approximation

4) Check that density is set properly

The material model itself is less likely to give problems
unless you have softening behaviour.

Trust this helps

Best Regards,
Nicolò Grilli

**Reply by Traiwit** [reply]
Thanks @ngrilli I will give it a go tmr morning

Just wondering is it possible to use use_displaced_mesh = false with a large strain problem? and where should I activate this option? I see it can be placed at most of the [Materials] block.

Thank you

Kind regards,
Traiwit

**Reply by bwspenc** [reply]
Currently, you have to use the displaced mesh for finite strain. However, you may be interested in this PR that will allow running without it: https://github.com/idaholab/moose/pull/17475

Just curious -- is the displaced mesh causing you trouble?

**Reply by Traiwit** [reply]
@bwspenc I don't think so, I'm just trying all the possible ways to make my simulation works. I am now back to implicit dynamic and trying to work around the plasticity.

Thanks anyway!

**Reply by bwspenc** [reply]
If you're having trouble with plasticity, I would recommend looking at using the automatic differentiation (AD) versions of those models. We know that our Jacobians aren't great with the non-AD versions of the plasticity models, which is surely going to hurt convergence.
